### DISCUSSION TITLE
AppCenter retirement - Migration guides

### DISCUSSION THREAD

**Discussion by bruno-garcia** [getsentry/sentry-dotnet#3226] [General]
With [AppCenter announcing its retirement](https://learn.microsoft.com/en-us/appcenter/retirement). I imagine more requests for migration guides will come.

Sentry supports **all platforms** that AppCenter supports + a lot more. So we should be able to help folks migrate their apps in first, and they'll be exposed to guides to add new apps (e.g: backend) that AppCenter didn't support yet.

https://nugettrends.com/packages?months=72&ids=Sentry&ids=Microsoft.AppCenter

![image](https://github.com/getsentry/sentry-dotnet/assets/1633368/9462ab36-bb83-4513-8fc3-f8a5cfe1f782)

### MAUI and Native

.NET 7 and newer has Native mobile supported by the [Sentry NuGet package](https://www.nuget.org/packages/Sentry/)
.NET MAUI is supported by [Sentry.Maui](https://www.nuget.org/packages/Sentry.Maui) and includes bindings to native iOS and Android libraries. These are super mature native SDK by Sentry ([sentry-android](https://github.com/getsentry/sentry-java) and [sentry-cocoa](https://github.com/getsentry/sentry-cocoa))

Sentry also supports **line numbers** on exceptions in release builds. ([which is still an open ticket on the appcenter GitHub repo](https://github.com/microsoft/appcenter/issues/165)). Symbols are uploaded automatically via [MSBuilds (docs)](https://docs.sentry.io/platforms/dotnet/configuration/msbuild/)

### Xamarin 

Sentry supports Xamarin (native + Forms):

* [Code on GitHub](https://github.com/getsentry/sentry-xamarin)
* NuGet packages ([Native](https://www.nuget.org/packages/Sentry.Xamarin), [Xamarin.Forms](https://www.nuget.org/packages/Sentry.Xamarin.Forms))

## WinForms

Through the [Sentry](https://www.nuget.org/packages/Sentry/) package. [Documentation here](https://docs.sentry.io/platforms/dotnet/guides/winforms/).

## WPF

Through the [Sentry](https://www.nuget.org/packages/Sentry/) package. [Documentation here](https://docs.sentry.io/platforms/dotnet/guides/wpf/).

## UWP & WinUI

Through the [Sentry](https://www.nuget.org/packages/Sentry/) package. [Documentation here](https://docs.sentry.io/platforms/dotnet/guides/winui/).

## Unity

Unity SDK [is on GitHub](https://github.com/getsentry/sentry-unity/). The package itself is published [via UPM](https://github.com/getsentry/unity). [Documentation is here](https://docs.sentry.io/platforms/unity/). 

### NuGet Trends

.NET Foundation project [NuGet Trends](https://nugettrends.com/) uses Sentry. [Code on GitHub](https://github.com/dotnet/nuget-trends).

### Symbol Collector

Sentry has a symbol collection service build in C# with an [Android app, CLI, and backend in ASP.NET Core that might be a reference for the API](https://github.com/getsentry/symbol-collector).

**Comment by Sevenish** [comment]
Just started using Sentry coming from AppCenter. One of the main questions that came up was how to handle additional data with an exception:
`// Exception ex, IDictionary<string, string> properties`
`Microsoft.AppCenter.Crashes.Crashes.TrackError(ex, properties)`

In the sentry documentation I find the concepts of breadcrumbs and contexts. But the data doesn't feel like an event (breadcrumbs) and context feel a bit broad (not specific for an exception).

After looking at the methods I found the most simple was to add an extra AddBreadcrumb call just before capturing the exception:
```
if (properties is not null)
{
  SentrySdk.AddBreadcrumb("exception_props", data: properties);
}    
SentrySdk.CaptureException(ex);
```
The AddBreadcrumb accepts IDictionary<string, string> which my client application uses allot when reporting exceptions. I suspect this will work fine. Is this a good approach? Addressing this specifically in a migration guide could help other developers coming from AppCenter.

**Reply by jamescrosswell** [reply]
@Sevenish it depends a bit on the nature of the data. If you want to be able to search by the data, then you could use [Tags](https://docs.sentry.io/platforms/dotnet/enriching-events/tags/). Otherwise I think a Context or a [Structured Context](https://docs.sentry.io/platforms/dotnet/enriching-events/context/#structured-context) are your best bet. 

The .NET SDK still supports adding arbitrary additional data via the SetExtra method but you can see from [the docs](https://docs.sentry.io/platforms/dotnet/enriching-events/context/#additional-data) that this will be phased out eventually.

**Reply by Sevenish** [reply]
Yes I see that Structured Context is a better long term approach. But I have a large code base that uses IDictionary<string, string> all over the place to give extra, unstructured context with errors. After testing I have concluded tha AddBreadcrumb is the simplest method without needing to convert the data and it replicates the functionally AppCenter has in the simplest and most straightforward way. In the meantime I can think about specific structured context objects for different error scenarios and gradually move towards using that.

**Reply by jamescrosswell** [reply]
OK... one thing to watch out for with Breadcrumbs is that the SDK will only keep a certain maximum number of these as determined by the [SentryOptions.MaxBreadcrumbs](https://docs.sentry.io/platforms/dotnet/configuration/options/#max-breadcrumbs) setting. If you're setting a breadcrumb immediately before an exception gets captured, this shouldn't be a problem... but if you add a breadcrumb and then capture an event (like an exception) only after some delay, it's possible the breadcrumb will get thrown away (and replaced by more recent breadcrumbs) before the capturing of that event takes place. 

Just something to be aware of (and you can increase the maximum number of breadcrumbs to reduce the chances of this happening).
